ESQL: Load `keyword` fields missing doc_values

csoulios opened the pull request:
This PR adds support for loading `keyword` fields having `doc_values: false` in their mapping. 

The code follows the same code path as when loading `text` fields, loading field from the index (if `stored` is `true`) or  extracting from the `_source`.

mark-vieira committed:
Add forcibly argument to test cluster stop node method (#98336)

(cherry picked from commit 2975f3379d8df50822f13489211393653b0462eb)

javanna committed:
Introduce executor for concurrent search (#98204)

This commit enables concurrent search execution in the DFS phase, which is going to improve resource usage as well as performance of knn queries which benefit from both concurrent rewrite and collection.

We will enable concurrent execution for the query phase in a subsequent commit. While this commit does not introduce parallelism for the query phase, it introduces offloading sequential computation to the newly introduced executor. This is true both for situations where a single slice needs to be searched, as well as scenarios where a specific request does not support concurrency (currently only DFS phase does regardless of the request). Sequential collection is not offloaded only if the request includes aggregations that don't support offloading: composite, nested and cardinality as their post collection method must be executed in the same thread as the collection or we'll trip a lucene assertion that verifies that doc_values are pulled and consumed from the same thread.

## Technical details

This commit introduces a secondary executor, used exclusively to execute the concurrent bits of search. The search threads are still the ones that coordinate the search (where the caller search will originate from), but the actual work will be offloaded to the newly introduced executor.

We are offloading not only parallel execution but also sequential execution, to make the workload more predictable, as it would be surprising to have bits of search executed in either of the two thread pools. Also, that would introduce the possibility to suddenly run a higher amount of heavy operations overall (some in the caller thread and some in the separate threads), which could overload the system as well as make sizing of thread pools more difficult.

Note that fetch, together with other actions,  is still executed in the search thread pool. This commit does not make the search thread pool merely a coordinating only thread pool, It does so only for what concerns the IndexSearcher#search operation itself, which is though a big portion of the different phases of search API execution.

Given that the searcher blocks waiting for all tasks to be completed, we take a simple approach of introducing a thread pool executor that has the same size as the existing search thread pool but relies on an unbounded queue. This simplifies handling of thread pool queue and rejections. In fact, we'd like to guarantee that the secondary thread pool won't reject, and delegate queuing entirely to the search thread pool which is the entry point for every search operation anyway. The principle behind this is that if you got a slot in the search thread pool, you should be able to complete your search, and rather quickly.

As part of this commit we are also introducing the ability to cancel tasks that have not started yet, so that if any task throws an exception, other tasks are prevented from starting needless computation.

Relates to #80693
Relates to #90700

original-brownbear committed:
Remove unnecessary locking in SparseFileTracker (#98315)

These callbacks don't need to synchronize outside of assertions since the
listener field is final -> remove synchronization and move assertion on
progress to a separate method. This might help performance a little,
since the onProgress callback can become rather hot during cache
writing.

martijnvg committed:
tidy up post collection. (#98345)

DaveCTurner committed:
Minor CoordinatorTests logging fixes (#98342)

Adds a little more detail to `ElectionSchedulerFactory`, and fixes the
wrapping of tasks in `DisruptableClusterApplierService` so that their
logs are decorated with the details of the owning node.

romseygeek committed:
Add an FS repository that simulates network download latencies (#98164)

This commit adds a new repository type `latency-simulating` that
duplicates the behaviour of an fs repository but adds a configurable
latency to blobstore  reads, simulating the time that a call to a remote
object store would take.

masseyke committed:
Hiding data stream lifecycle documentation in released docs (#98334)

abdonpijpelink committed:
[DOCS] Update CCS compatibility matrix for 8.10 (#98341)

ioanatia committed:
Search Applications: Remove inconsistent indices warning (#98348)

* Remove inconsistent indices warning

* Revert docs changes for inconsistent indices warning

DaveCTurner committed:
Suppress log for reroute failure on stopping master (#98278)

If the master stands down between a put-shutdown update and the
subsequent reroute then today we emit a `WARN` log, but this is benign
(and possibly expected) so with this commit we suppress the warning.

carlosdelest committed:
Synonyms Overview Documentation (#98202)

mark-vieira committed:
Simplify extending cluster handle implementations (#98363)

jrodewig committed:
[DOCS] Force merge doesn't mark index as read-only (#98357) (#98366)

(cherry picked from commit d4c6208727fe0b76906006fd699e7ac5bf9474f6)

Co-authored-by: Gilad Gal <gilad.gal@elastic.co>

tvernum committed:
Allow rest tests with Serverless API protections (#98250)

This commit changes the `ESRestTestCase` in 3 ways - Refactor the
"hasXYZ" flags into a single `EnumSet` of `ProductFeature` flags -
Detect serverless distributions, and in those cases treat certain
features as unavailable (even if the plugin is installed) - Add a few
more feature checks for features that can now be disabled

ywangd committed:
Bump TransportVersion (#98368)

Bump TransportVersion for stateless

tvernum committed:
Revert "Allow rest tests with Serverless API protections (#98250)" (#98373)

This reverts commit 2ba7b798ad85d1e21883eb7a5ff26596a2d66c41.

Temporarily reverting because it breaks downstream projects

ywangd committed:
[Test mute] AwaitsFix #98372

AwaitsFix #98372

DaveCTurner committed:
More cleanup of TransportAddVotingConfigExclusionsActionTests (#98356)

Rather than waiting for both cluster state updates, we should assert
that the second update has executed, performing the reconfiguration, by
the time the API returns.

gmarouli committed:
Add data stream lifecycle by default (#97823)

In this PR we enable all new data streams to be managed by the data
stream lifecycle by default. This is implemented by adding an empty
`lifecycle: {}` upon new data stream creation. 

Opting out is represented by a the `enabled` flag:

```
{
  "lifecycle": {
    "enabled": false
  }
}
```

This change has the following implications on when is an index managed
and by which feature:

| Parent data stream lifecycle| ILM| `prefer_ilm`|Managed by|
|----------------------------|----|----------------|-| | default | yes|
true| ILM| | default | yes| false| data stream lifecycle| |default |
no|true/false|data stream lifecycle| |opt-out or
missing|yes|true/false|ILM| |opt-out or missing|no|true/false|unmanaged|

Data streams that have been created before the data stream lifecycle is
enabled will not have the default lifecycle.

Next steps: - We need to document this when the feature will be GA
(https://github.com/elastic/elasticsearch/issues/97973).

martijnvg committed:
Remove the usage of rollup from the rollup qa module, (#98387)

which is actually only testing downsampling.
Also rename a few classes / directories to be be downsample oriented.

carlosdelest committed:
Rename Synonyms API namespace (#98383)

felixbarny committed:
[DOCS] ILM force merge action doesn't make index read-only (#98382)

abdonpijpelink committed:
[DOCS] Set up temporary redirects for new remote clusters docs (#98398)

jimczi committed:
Add the total dense vector count in the indices stats output (#98275)

This change adds the total dense vector count to the output of the indices stats.
This is useful for observability in order to track the number of indexed vectors
in a cluster.

---------

Co-authored-by: Benjamin Trent <ben.w.trent@gmail.com>

original-brownbear committed:
Remove array lookup indirection from read-path in SharedBytes (#98399)

There's no real reason to have the indirection of tracking free regions
by integer and using `-1` as a `null` of sorts and then looking regions up
from an array.
=> refactored the logic to just hold a reference to the IO region
directly to save some volatile reads and array lookups. We can simplify
this further in a subsequent step by removing the complexity of passing
the region start offset into the `read` method and then redundantly
subtracting it there.

albertzaharovits committed:
Tweak security system indexes refresh behavior for stateless (#98285)

This is an omnibus PR that:
 * toggles the fast_refresh option for the profiles index (to make it alike to the .kibana index)
 * makes API Key creation refresh policy default to IMMEDIATE from WAIT_UNTIL in serverless, to mitigate the long automatic refresh interval

masseyke committed:
Fixing data stream lifecycle documentation links that break clients (#98365)

brianseeders committed:
Add automation for building/uploading/updating lucene snapshots daily (#98039)

quux00 committed:
Log CCS errors at DEBUG level (#98290)

* Log CCS errors at INFO level

In order to gather data on what types of CCS errors happen in the field, we will log
them using the ShardSearchFailure XContent (JSON), which supplies information about
underlying causes of shard failures.

* Change to WARN level

* Changed log levels to DEBUG

luigidellaquila committed:
ESQL: Add support for negative numbers in multivalues for ROW operator (#98359)

csoulios committed:
Load keyword fields when they have no doc_values

brianseeders committed:
[lucene_snapshot] Fix schedules, set default branch, and pass branch to triggered builds (#98410)

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc

brianseeders committed:
Fix lucene_snapshot cron definitions

masseyke committed:
Unmuting BootStrapTests.testMixedTriggeredWatchLoading() (#98407)

masseyke committed:
Fix assertions in GeoIpDownloaderIT.testGeoIpDatabasesDownloadNoGeoipProcessors() (#98413)

bpintea committed:
ESQL: Pushdown metadata attributes (#98355)

This allows to push down those metadata attributes that can be pushed
down.
Out of the currently supported three, the only one pushed down is
_index. This only supports wildcard and terms queries.

bpintea committed:
Verify the constant types assigned in a row (#98403)

This adds a verification for the constant types assigned directly into
a ROW source command.
Types like Duration and Period cannot be used, as they are not
representable in a block.

original-brownbear committed:
Remove dead exception handling code in InboundHandler (#98421)

The code inside the try never throws and there's no point in asserting that much,
the code obviously does not throw and we have exception handling that prevents leaks further up
the stack anyway in case of a bug.

ywangd committed:
Remove RCS 2.0 feature flag for beta release (#98307)

This PR removes the RCS 2.0 feature flag so that it is ready for beta
release.

quux00 committed:
Minor improvements for Cross-Cluster Search code (#98405)

Several minor structural and test improvements for cross-cluster search

These changes set the stage for a follow-on ticket to add _cluster/details to
cross-cluster searches with minimize_roundtrips = false. To help keep that
PR from being too large some of the simpler required changes and tests
are added in this PR.

tvernum committed:
Mark "_features/" API internal on Serverless (#98375)

This API may be useful as part of operational backups

tvernum committed:
YAML rest test for API keys without create user (#98429)

Adds a new YAML based rest test that verfies the majority of API key
functionality without creating any custom users or roles. This allows
it to be run in environments where native user management is not
available.

DaveCTurner committed:
Fix refcounting in action base classes (#98381)

Transport requests may hold nontrivial resources, and support
ref-counting to ensure prompt release of those resources. The base
classes that implement various delegating action patterns create new
requests, or retain references to requests after returning to the
caller, and must therefore integrate properly with the ref-counting
mechanism on those requests. The following classes do not do so today,
and this commit fixes that:

- `TransportBroadcastByNodeAction`
- `TransportMasterNodeAction`
- `TransportNodesAction`
- `TransportTasksAction`

DaveCTurner committed:
Make TransportAddVotingConfigExclusionsAction retryable (#98386)

The docs for this API say the following:

> If the API fails, you can safely retry it. Only a successful response
> guarantees that the node has been removed from the voting
> configuration and will not be reinstated.

Unfortunately this isn't true today: if the request adds no exclusions
then we do not wait before responding. This commit makes the API wait
until all exclusions are really applied.

ywangd committed:
Update changelog highlight for RCS 2.0 (#98427)

Highlight the feature

Relates: #98307

DaveCTurner committed:
Implement MockPersistedState#toString (#98431)

These things appear in logs & test failure messages, makes sense to show
something a little more useful than the default `toString()` impl.

thomasdullien committed:
Fixed scaling bug. (#98350)

* Fixed scaling bug.

* Fix tabs to spaces

---------

Co-authored-by: Tim Rühsen <tim.ruhsen@elastic.co>

ywangd committed:
[Test] Tweak RCS 2.0 tests (#98426)

General improvements for RCS 2.0 related tests.

ywangd committed:
Support configuring mTLS for RCS 2.0 (#98428)

RCS 2.0 uses server TLS by default but should support mTLS when needed.
This PR enables relevant settings for mTLS and adds a test suite for
mTLS setup.

original-brownbear committed:
Deduplicate DocIdSetIterator in ProfileScorer (#98418)

It's the same anonymous type here in both spots => dedup.
Also, removed some dead code `throws IOException` related to this code.

edsavage committed:
Mute failing testContinuousDateHistogramPivot IT (#98441)

Mute failing integration test:
`org.elasticsearch.xpack.transform.integration.TransformPivotRestIT.testContinuousDateHistogramPivot`.
See  https://gradle-enterprise.elastic.co/s/bcizcespvmuhu

Related to: #98377

ChrisHegarty committed:
Add unavailable EsqlFeatureSetUsage creation support (#98438)

This commit adds a means to create an EsqlFeatureSetUsage that is unavailable and not enabled. This required to be able to mock out alternative / dummy implementations of usage / info transport actions where the ESQL module is not present, thus returns empty / unavailable feature usage stats, etc.

This is a similar to other not-present xpack features in stateless, e.g. ILM.

DaveCTurner committed:
Preserve PeerFinder state when deactivated (#98434)

Today the `PeerFinder` releases all its peers on deactivation, but to
complete #98354 we will need to delay the release until the cluster has
reached a more stable state, which happens some time later. This commit
separates the deactivate and release steps within the `PeerFinder` in
preparation for that change, although it still does both steps at once
in all production callers.

DaveCTurner committed:
Fix election scheduling after discovery outage (#98420)

Adds a test showing that when an extended discovery outage eventually
clears up, an election can happen very promptly (and fixes the
calculation of `prevElectionWon` to avoid a situation where that would
not happen today). This demonstrates a subtle problem that the initial
version of #98354 would have introduced.

ChrisHegarty committed:
Remove space in dissect feature list

DaveCTurner committed:
Improve reliability of elections with message delays (#98354)

Today we close the election scheduler when the coordinator leaves mode
`CANDIDATE`, before even starting the publication that establishes the
election winner as the cluster master. If this publication subsequently
fails then we start a new election scheduler with the original, short,
timeout, and do not back off. With very high numbers of master-eligible
nodes this can lead to constant election clashes that never resolve. We
must count such failed publications as failed election attempts for
election scheduling and backoff purposes.

This commit keeps the election scheduler open until a published state is
applied, which means we continue to back off until a publication has
completed.

Closes #97909

DaveCTurner committed:
Remove write pool constraints from RejectionActionIT (#98433)

These tests are not trying to trigger rejections from the write
threadpool, so it would be best to have the default config here.

Closes #97291

javanna committed:
TimeSeriesIndexSearcher to offload to the provided executor (#98414)

We have introduced a search worker thread pool with #98204 that is
responsible for the heavy workloads as part of the query and dfs phase,
no matter if it is parallelized across segments/slices or not. TSDB
aggregations are still executed in the search thread pool and this
commit moves their computation to the search worker thread pool, despite
the corresponding search thread blocks and  waits for such computation to be
completed before returning.

javanna committed:
Work around adaptive offloading of knn rewrite tasks (#98408)

Lucene has a non configurable QueueSizeBasedExecutor that inspects the
executor, and uses the queue size compared to the max pool size to
decide whether to executor some of the parallelizable tasks on the
caller thread or on the executor. We'd rather want every I/O or CPU
bound task executed on the recently introduced search worker thread
pool.

We have already worked around this in our ContextIndexSearcher
customizations when it comes to the collection part of the search. The
rewrite part is more difficult to customize, but we can temporarily wrap the
executor to make it not an instanceof ThreadPoolExecutor, to disable
such adaptive mechanism that we don't wish to leverage.

masseyke committed:
Not checking index.time_series.end_time in 150_tsdb's `created the data stream` test because it changes (#98458)

AthenaEryma committed:
Check ILM status before reporting node migration STALLED (#98367)

This commit adjusts the behavior of the node shutdown status
calculation to avoid reporting a `STALLED` status when ILM is
in the process of doing something that might result in the
migration being temporarily stalled (at time of writing, that
means shrinking an index).

This check is only made if ILM is not stopped.

Leaf-Lin committed:
Update post-analytics-collection-event.asciidoc (#97826)

* Update post-analytics-collection-event.asciidoc

Update API endpoint. Instead of `click`, it should be `search_click`. 

Expanded on request body in the API as well.

* Update post-analytics-collection-event.asciidoc

* Update post-analytics-collection-event.asciidoc

* Update post-analytics-collection-event.asciidoc

tvernum committed:
Add YAML test for authentication without user API (#98443)

Adds a test for the security.authenticate (/_security/_authenticate)
API that doesn't depend on the native realm being available

Relates: #98429

tvernum committed:
Support path filtering in YAML test suite (#98430)

Some Rest Specs contain multiple paths, where some paths are
preferred and the others exist for historical reasons.

This commit makes 2 changes:

1. Tracks when a path is deprecated in the Rest Spec, so that metadata
   can be accessed in the Java model of the API
2. Adds filter (predicate) to the execution context to determine which
   paths should be considered as candidates within the test suite

Tests that extends from `ESClientYamlSuiteTestCase` can override
`createRestTestExecutionContext` and pass in a predicate that skips
deprecated paths (or any other criteria), provided that it does not
skip _all_ paths in an API.

salvatore-campagna committed:
Run downsampling using persistent tasks (#97557)

Add support for downsampling persistent tasks. We would like to have the ability
to resume downsampling tasks, instead of starting from scratch in case of
("retriable") failures.

Instead of keeping track of the state of the task we just try to query the downsampling
target index before starting the actual downsampling task, getting the latest document
indexed and its tsid. If we find any, we just restart from that tsid, maybe
overwriting a subset of the documents already indexed (documents with the same
tsid whose timestamp is smaller then the one of the latest document).

Querying the downsampling target index is possible after introducing a predictable
naming scheme for the target index. This was not the case before, a part of the target
index name was random. That is required to be able to start from where the task left.

Note that since the ordering of documents is based on (tsid, timestamp) we do not
include the timestamp in the query to avoid skipping downsampling of documents
with a larger tsid but smaller timestamp.

Note also that we are not removing the TransportDownsampleIndexerAction for
backward compatibility. An older master node running a previous version would
not use persistent tasks and an API call would result in something like "missing
downsample action".

We are also adding an additional downsample rest query parameter, timeout.
It allows a user to set the maximum time taken while waiting for a downsampling
task to complete before returning with an timeout error. Note that when the timeout
triggers the task will still run under control of the executor, just waiting for it to finish
will result in a timeout. The default value for this timeout is 1 day.

Closes #93582

---------

Co-authored-by: Martijn van Groningen <martijn.v.groningen@gmail.com>

tvernum committed:
Make GET /_nodes/shutdown internal on serverless (#98424)

On serverless we don't use Rest APIs to create shutdown events, but we
do want to checkt the status of an event

martijnvg committed:
Move downsample feature to its own xpack plugin (#98422)

Due the fact that downsampling and rollup where in the same module, it is sometime confusing what is what and rollup and downsample terminology are easily mixed up. Splitting downsampling in its own x-pack module should reduce confusion. There are some classes, fields, methods, variables and comments that need renaming, but this PR is focussed around moving downsampling to its module and renames would make it more difficult to review. I will do renaming in a follow up PR.

The upside is that rollup no longer needs to depends on the aggregate double module, only downsampling depends on this and the downsample plugin doesn't need all the transport and rest actions and scheduler service that rollup depends on.

javanna committed:
[TEST] Remove ExecutionException handling from AggregatorTestCase (#98415)

ExecutionException is no longer thrown by the searcher: runtime
exceptions (like CircuitBreakingException) are re-thrown unchanged,
so the catch needed for concurrent search is the same as the one for
sequential search.

javanna committed:
Unwrap IOException in ContextIndexSearcher concurrent code-path (#98459)

The search method that handles parallel collection across segment slices
catches and unwrap execution exception that each slice may throw. It
will rethrow the cause as-is if it's a runtime exception, otherwise wrap
it into a new runtime exception before re-throwing. Given that the
method itself throws IOException, we should refrain from wrapping
IOExceptions with a new runtime exception.

abdonpijpelink committed:
[DOCS] Add tests to behavioral analytics code snippets (#98378)

pierrehilbert committed:
Index mapping: components fields is now mapped as nested with two properties as keywords. Will be use for telemetry collection. (#98471)

thecoop committed:
Remove link between Version and IndexVersion (#97513)

This removes the code link between Version and IndexVersion, so they are now independent classes

DaveCTurner committed:
Clean up handling of TransportResponse.Empty (#98447)

There's no need for `TransportResponseHandler.Empty` to pass the empty
instance into its `handleResponse()` method, and there's no need for
`EmptyTransportResponseHandler` at all. Also cleans up a few places
where `TransportResponse.Empty` appears in APIs that should just be
using `Void`.

andreidan committed:
Ensure frozen indices have correct tier preference (#97967)

We've seen a series of cases where partially mounted indices either had
an invalid tier preferece (e.g. `data_content`) or a missing tier
preference.

This tightens up our safety nets to make sure that when we change the
index configuration (e.g. we remove the tier preference using the update
settings API), migrate a deployment to data tiers, restore an index, or upgrade
the persisted metadata we set the tier preference to `data_frozen` for
partially mounted indices.

martijnvg committed:
Mute two DownsampleClusterDisruptionIT tests (#98486)

Relates to #98485

andreidan committed:
Make sure the downsample step is not stuck (#98477)

If the downsampling index exists but it's still downsampling we should
invocate the downsample transport action again (and wire up the request
listener such that the ILM listener gets notified of success or failure)

This modifies the test to assert the ILM listener is invoked and removes
an invariant that doesn't hold anymore in ILM (i.e. previously before
#97557 but now we can, and do)

Markin as non-issue as this hasn't been released yet.

martijnvg committed:
Generate a correct consistent downsample target name in downsample ilm action (#98476)

Bring back generate downsample step that is used to prepare the
lifecycle state and sets the name of the target downsample index to be
used. A number of steps will use this to execute various downsample
related operations.

The generate downsample index name step produces a consistent name. And
also takes into account whether the source index was already
downsampled, so a logical name is generated.

Relates to #93582

(marking as a non issue, because the bug isn't part of an released
version of Elasticsearch)

DaveCTurner committed:
Revert "Generate a correct consistent downsample target name in downsample ilm action (#98476)"

This reverts commit 8676b8c752ea1acddcf312c08d161f2206213a06.

javanna committed:
Add support for concurrent collection when size is greater than zero (#98425)

The query phase supports so far concurrency only when size is set to 0 (despite inter-segment concurrency is not yet enabled in the query phase). This commit adds support for concurrent collector managers to the query phase when size is greater than zero (hits are requested), including when scroll is used. The only exception is field collapsing, which relies on custom collectors for which we don't have yet a corresponding collector manager.

Note: this commit does not enable concurrent collection for the query phase. The entire query phase still collects against a single slice, sequentially segment by segment.

Collector managers refactoring / suppliers removal
As part of this change, especially figuring out how to leverage the lucene collector managers for top score doc and top field doc collectors, many simplifications were made, that are explained as follows.

We currently create multiple collector managers for the query phase: one for the top docs collection, one for aggs, and both of them are wrapped into a query phase collector manager. Additionally, each one of those is wrapped into a ProfileCollectorManager if profiling is enabled. While this works, the 1:1 ratio between collectors and collector managers is not necessary and makes things quite hard to follow.

Also, TopDocsCollectorManagerFactory used to be a factory to create the right top docs collector based on the requested features, and it is now a factory of collector managers. We can simplify things and remove one level of indirection by converting it into a collector manager directly. Also, we use some suppliers to keep track of which collector provides what part of the search response (e.g. total hits, top docs, max score). This is because of the collectors wrapping: before the introduction of collector managers, it was difficult to extract the appropriate info from each collector after calling search, problem that was solved using suppliers. Collector managers are a good fit to replace these indirections, as they can return a value, which is the result of the reduction of the results obtained by each slice after collection.

This commit folds all of the different collector managers that we rely on for the query phase into a single one (called QueryPhaseCollectorManager): its newCollector method will create the appropriate collectors depending on the requested features (a bit like TopDocsCollectorManagerFactory used to do), but it will take care of the creation of all the collectors involved, depending on feature requested (size, scroll, collapse) or whether profiling was enabled. This allows us to have a single base collector manager that includes all the logic around creating collectors and reducing their results, as well as returning a concrete result which includes the top docs, terminated after flag, profiling tree if profiling was enabled which entirely removes the need for suppliers in multiple places.

There are no suppliers needed for total hits and top docs, as they are returned directly by the search method. There is no need for collector managers to hold state that is only set once reduce is called. Profiling was greatly simplified as well to no longer rely on suppliers for the profiler results. Instead, the collector results can simply be set to the query profiler itself after search returns. Additionally, ProfileCollectorManager used to have handling for children collectors for top docs and aggs, which was very hard to follow. This is no longer required because QueryPhaseCollectorManager takes care of it all (wrapping top docs, aggs, as well as the top level collector when profiling is enabled) and removes the need for wrapping of collector managers.

thecoop committed:
Bulk migration of Version.CURRENT for index created to IndexVersion.current() (#98490)

masseyke committed:
Changing watcher to disable cookies in shared http client (#97591)

abdonpijpelink committed:
[DOCS] Update geohash_grid agg field description (#98494)

DaveCTurner committed:
Improve election scheduling tests with atomic-register coordinator (#98491)

We can at least verify that the election scheduler is not running on
nodes that have a healthy connection to the register.

Relates #98423

martijnvg committed:
Revert the commit that reverted the 'Generate a correct consistent downsample target name in downsample ilm action' change (#98496)

The revert the commit (504aa401af723fb340a4abbcc8afd43130538c96) that
reverted #98476

fcofdez committed:
Allow responses in TransportBroadcastUnpromotableActions (#98445)

Relates ES-5061

davidkyle committed:
[ML] Change logging level used in NLP request processing (#98497)

fcofdez committed:
Bump TransportVersion (#98500)

amyjtechwriter committed:
[DOCS]Corrections for index-mgmt page. (#98481)

* Corrections for index-mgmt page.

* Update docs/reference/indices/index-mgmt.asciidoc

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

---------

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

thecoop committed:
Revert "Remove link between Version and IndexVersion (#97513)" (#98501)

This reverts the change to separate IndexVersion, so that nothing is modified for the 8.10 release. When 8.10 is branched, we can then re-apply this change so that all the changes to IndexVersion happen for 8.11, simplifying the build infrastructure for these releases.

saikatsarkar056 committed:
[Search Relevance]  Refine the license compliance error message to be specific for features (#98409)

* Specify the feature in the exception message

* Specify the feature in the exception message

* Specify the feature in the exception message

* Specify the feature in the exception message

* Specify the feature in the exception message

* Specify the feature in the exception message

* Specify the feature in the exception message

DaveCTurner committed:
Remove unused verifyElectionSchedulerState() (#98499)

Closes #98423

jakelandis committed:
Fix jwt and oidc doc (#97942)

This commit corrects the following issues with JWT and OIDC `jwkset_path` documentation:
* only https is supported for the JWT realm (OIDC support both https and http)
* JWT realm does not use a file watcher to reload the file every 5 seconds
* simplify "path" to  "file name" ..technically it is resolved path, but 99% of the time it will be just 
     a file name in the config directory and "path" is ambiguous
* remove special mention of using the absolute path in cloud. .. this is an unnecessary implementation 
     detail and the only setting (of many) that calls out the cloud config directly by absolute path
* ensure the 2 different JWT documentations are the same
* make mention of when the JWT file will be reloaded (it is not backed by the file watcher, only OIDC is)

nik9000 committed:
ESQL: Support `null` group key in more cases (#98507)

This adds support for null group keys in all groupings that are
backed by the PackedValuesBlockHash. This means the only groupings
that don't support null keys at the moment are long/long pairs,
and long/bytes_ref pairs. Those are coming next.

nik9000 committed:
ESQL: Optimize some `MV_` functions on sorted (#98515)

When multivalued fields are loaded from lucene they are in sorted order
but we weren't taking advantage of that fact. Now we are! It's much
faster, even for fast operations like `mv_min`

```
     (operation)  Mode  Cnt  Score   Error  Units
          mv_min  avgt    7  3.820 ± 0.070  ns/op
mv_min_ascending  avgt    7  1.979 ± 0.130  ns/op
```

We still have code to run in non-sorted mode because conversion functions
and a few other things don't load in sorted order.

I've also ported expanded the parameterized tests for the `MV_` functions
because, well, I needed to expand them at least a little to test this
change. And I just kept going and improved as many tests as I could.

volodk85 committed:
Handle ES internal assertion errors in IT test (#98517)

Indexing test thread throws `AssertionError` which is not handled, so it
escapes to main **test thread**

```
aug 14, 2023 10:39:19 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException |  
-- | --
  | WARNING: Uncaught exception in thread: Thread[elasticsearch[node_s0][write][T#4],5,TGRP-BulkIntegrationIT] |  
  | java.lang.AssertionError: primary indexing unexpected in state [CLOSED] |  
  | at __randomizedtesting.SeedInfo.seed([8C8A69C029D945DF]:0) |  
  | at org.elasticsearch.index.shard.IndexShard.assertStartedForPrimaryIndexing(IndexShard.java:2204) |  
  | at org.elasticsearch.index.shard.IndexShard.ensureWriteAllowed(IndexShard.java:2184)
```

full test
[log](https://gradle-enterprise.elastic.co/s/oraopkwtuyble/tests/task/:server:internalClusterTest/details/org.elasticsearch.action.bulk.BulkIntegrationIT/testDeleteIndexWhileIndexing?top-execution=1)

uncaught exception releases `thread.join` wait and **test thread** might
be able to execute following `assertFalse` statement. Indexing thread
might be technically alive by the time of `assertFalse` execution (e.g.
due to jvm housekeeping so that thread status is not reported to OS in
time thus **isAlive**  as native method still thinks that thread is ok)

```
            thread.join(ReplicationRequest.DEFAULT_TIMEOUT.millis() / 2);
            assertFalse(thread.isAlive());
```

Closes #98435

ywangd committed:
[Docs] Misc doc update for RCS 2.0 (#98472)

This PR adds docs for the following items: * Remote indices privileges *
Remote cluster network settings * Remote cluster security settings * New
privileges * New response field for RemoteInfo API

List of preview pages: * [Remote indices in defining
roles](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/defining-roles.html#roles-remote-indices-priv)
* [Remote indices in PutRole
API](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-api-put-role.html#security-api-put-role-request-body)
* [Remote cluster server SSL
settings](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-settings.html#_remote_cluster_server_api_key_based_model_tlsssl_settings)
* [Remote cluster client SSL
settings](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-settings.html#_remote_cluster_client_api_key_based_model_tlsssl_settings)
* [Remote cluster network
settings](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/modules-network.html#remote-cluster-network-settings)
and
[here](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/modules-network.html#common-network-settings)
* [Remote cluster credentials
setting](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/remote-clusters-settings.html)
* [New
privileges](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-privileges.html)
* [New response field for RemoteInfo
API](https://elasticsearch_98472.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/cluster-remote-info.html#cluster-remote-info-api-response-body)

Commit:
Mapped field types searchable with doc values (#97724)

* Mapped field types searchable with doc values

When using TSDB, time series metrics aren't indexed but do have doc values.
Field caps should report those fields as searchable.

jbaiera committed:
Allow custom geo ip database files to be downloaded (#97850)

This PR extends the assumptions we make about database file availability to all database file 
names instead of the default ones we host at Elastic. When creating a geo ip processor with 
a database name that is not recognized we unilaterally convert the processor to one that 
tags documents with a missing database message until the database file requested is 
downloaded or provided via the manual configuration route. This allows a pipeline to be 
created and for the download service to be started, potentially sourcing the needed files.

---------

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

DaveCTurner committed:
Revert "Handle ES internal assertion errors in IT test (#98517)"

This reverts commit 5199a7a3a294577c4dae5e55b921de34b06a8576.

martijnvg committed:
Tweaked ILMDownsampleDisruptionIT
by adding a client side timeout.

ChrisHegarty committed:
Update docs/changelog/98309.yaml

ChrisHegarty committed:
Post-merge fix - update TransportResponseHandler.Empty::handleResponse

JVerwolf committed:
Add a setting to conditionally enable concurrency during the query phase  (#98455)

This commit adds a cluster setting that allows to turn on inter-segment search concurrency in the query phase. The default value is false.

DaveCTurner committed:
Fix assertStartedForPrimaryIndexing (#98519)

Primary indexing is expected in states STARTED and CLOSED but this
assertion was only permitting it when STARTED. This commit fixes that
and also extracts the whole sequence of assertions to a separate method.

ChrisHegarty committed:
Fix changelog area

ChrisHegarty committed:
Post-merge fix - update TransportResponseHandler.Empty::handleResponse (again)

DaveCTurner committed:
Move instructions about new transport versions (#98078)

Today the instructions about adding a new transport version are in the
class javadoc, but these docs are really more for the users of a class
than its maintainers, and it's easy to miss that there are vital
instructions here.

This commit moves them somewhere a little less missable, and clarifies
that the list of versions should be _only_ versions, no comments or
whitespace.

romseygeek committed:
Make AsyncSearchResponse use Chunked ToXContent streaming (#98380)

Relates to #95661

martijnvg committed:
Rename many rollup usages in downsample codebase. (#98469)

A relative large change, but mainly this renames comments, variables, methods, fields and classes with rollup in the name to downsample. Also a few classes from xpack core module got moved from (org.elasticsearch.xpack.core.rollup to org.elasticsearch.xpack.core.downsample).

ChrisHegarty committed:
Mute SearchServiceTests::testIsParallelCollectionSupportedForResults

kkrik-es committed:
`Sqrt` function for ESQL (#98449)

* Sqrt function for ESQL

Introduces a unary scalar function for square root, which is a thin
wrapper over the Java.Math implementation.

* Fix area for ESQL integration changelog.

* Restore changelog.

* Restore area in changelog.

masseyke committed:
Adding special logic to the disk health check for search-only nodes (#98508)

This prevents search-only nodes from triggering a YELLOW health result
when they use more than 90% of a disk.

thecoop committed:
Use compatible node version for the test rollup version rather than index version (#98474)

thecoop committed:
Modify upgrade test version check to include 8.10 (#98536)

IndexVersion hasn't yet been separated from Version. So it's ok to modify this check to allow 8.10 releases to run the test.

alex-spies committed:
ESQL: Enable expressions with arithmetic negation (#98352)

Enable unary arithmetic negations in expressions, like eval x = -y.
Support integer, long and double arguments.

Changes:
* Add mapper for Neg expressions.
* Add tests.
* Disallow negating unsigned longs during query verification.

andreidan committed:
Move DownsampleAction and DownsampleTask to server (#98533)

This moves the downsample action and task to server so we can access
them from `modules`.

quux00 committed:
Bump to version 8.11.0

ldematte committed:
Mute MultiVersionRepositoryAccessIT tests (#98440)

* Re-enable suppressed tests to run them through CI

* Change AwaitsFix notation to reflect the current tracking issue

saikatsarkar056 committed:
Delete extra space from exception message (#98516)

original-brownbear committed:
Refactor InboundHandler into smaller Methods (#98444)

I tried to go through this code with someone and it became evident that
the massive method for request handling is very hard to grok. I split it
up into its logical constituents here and fixed some redundant passing
around of `message` and `header`. This at least makes it so we only have
one layer of ref count handling per method and it is a little easier to
follow where forking happens. Also, this removes a weirdly long
multi-second compile task from C2 for whatever that's worth :)

Also: * fixed `assertRemoteVersion` to be a proper assertion method *
extracted `verifyRequestReadFully` since it's not essential to the logic
just like we have `verifyResponseReadFully`

astefan committed:
ESQL: replace the is_null function with IS NULL and IS NOT NULL predicates (#98412)

nik9000 reviewed (commented):
I think it'd be nice to make these pluggable so types not in server could handle ESQL on their own. But that's a problem for another time. 

But as it stands this seems fine.

The trouble is that some of these

Commit:
Documentation update that fixes a query rules code example (#98540)

* Change example field in rule query guide

* Change fuzzy to contains to get tests to work

---------

Co-authored-by: Kathleen DeRusso <kathleen.derusso@elastic.co>

costin committed:
Prune unused columns from the query (#98462)

Remove unused aggregations or eval fields that are not needed in the
 final result.
For example the query:

from employees
| stats c = count(salary), min = min(salary)
| eval x = emp_no
| keep c

is now optimized as:

from employees
| stats c = count(salary)
| keep c

since neither the min or nor the x fields are actually needed.

tvernum committed:
Add YAML tests for Authorization using API Keys (#98527)

Adds 2 new YAML tests that test authorization behaviour exclusively
using API Keys

1. Indexing
2. Field Level Security

tvernum committed:
Mute JobNodeSelectorTests (#98554)

Relates: #98552

williamrandolph committed:
Add mappings version number to SystemIndexDescriptor (#97934)

This PR adds a mappings version number to system index descriptor, which is intended to replace the use of Version to signal changes in the mappings for the descriptor. This value is required for managed system indices.

Previously, most of the system index descriptors automatically incremented their mapping version with each release, which meant that the mappings would stay up-to-date with any additive changes. Now, developers will need to increment the mapping version whenever there's a change.

* Add MappingsVersion inner class to SystemIndexDescriptor
* Add mappings version to metadata in all system index mappings
* Rename version meta key ('system' -> 'managed')
* Update mappings for ML indices if required
* Trigger ML index mappings updates based on new index mappings version.

---------

Co-authored-by: Ed Savage <ed.savage@elastic.co>

tvernum committed:
Mute upgrades.FullClusterRestartIT.testClusterState (#98556)

Relates: #98555

tvernum committed:
Mute testCCSCheckCompatibility (#98559)

These tests fail when `TransportVersion.current()` is equal to
`TransportVersion.MINIMUM_CCS_VERSION`

Relates: #98557

tvernum committed:
Mute MlJobSnapshotUpgradeIT (#98561)

Relates: #98560

tvernum committed:
Mute FeatureUpgradeIT (#98563)

Relates: #98562

tvernum committed:
Mute TransformPersistentTasksExecutorTests (#98565)

Relates: #98564

thecoop committed:
Specify correct current IndexVersion after 8.10 release (#98574)

The current IndexVersion wasn't bumped by the release automation as it should have been

thecoop committed:
Fix tests that break when minimum ccs compatibility == current transport version (#98571)

Add a method to get one higher than minimum ccs version, even if the version hasn't been defined yet

thecoop committed:
Apply spotless properly

edsavage committed:
[ML] Mute failing TransformConfigUpdateTests test (#98579)

Relates #98577

thecoop committed:
Update allowed inferred versions to include 8.10 (#98578)

Update some tests that could pass CURRENT to infer versions to real versions

thecoop committed:
Fix shard-stores doc test (#98581)

The replacement wasn't actually doing anything due to different JSON types

thecoop committed:
Fix another CCS compatibility test (#98585)

thecoop committed:
Unmute FeatureUpgradeIT test (#98576)

Test was fixed by #98574

edsavage committed:
[M]] Fix failing JobNodeSelectorTests (#98541)

Tests have started failing in 8.11 as they are missing an appropriate value for the ML_CONFIG_VERSION_NODE_ATTR node attribute

edsavage committed:
[ML] Fix failing TransformConfigUpdateTests test (#98582)

Use TransformConfigVersion string throughout.

Relates #98577

edsavage committed:
[ML] fix failing TransformPersistentTasksExecutorTests (#98575)

Update test nodes with an appropriate value for the TRANSFORM_CONFIG_VERSION_NODE_ATTR attribute

Relates https://github.com/elastic/elasticsearch/issues/98564

craigtaverner committed:
Simple grammar fix for MVT docs (#98591)

martijnvg committed:
Fix downsample failure when FLS/DLS is enabled. (#98587)

If FLS/DLS is enabled (this is the case when trial/licence is active and
security is enabled) then invoking the downsample API results in
immediate failure.

The downsample shard persistent task executor opens a searcher, but
security didn't set indices permissions in the thread local (this
happens via SecurityActionFilter). This will only happen on indices
actions (which are actions with a request that implement IndicesRequest.
This change does this by delegating to a transport action that executes
always locally, and this way security prepares thread local headers
correctly.

This adds another layer of indirection, but without doing this FLS/DLS
wouldn't work.

Closes #98569

(Marking as a non-issue, because this a bug in not released code)

carlosdelest committed:
Dense vector field types are indexed by default (#98268)

* First version

* Spotless, I liked my version better

* Fix param default values

* Add a supplier for default value to ensure it's calculated correctly

* Can't improve this without breaking tests

* Added checks for not specifying a body in PUT requests

* Fix default provider for enum params

* Added yaml test

* Changed docs and fix TODO

* Removing synonyms changes

* Added separate methods for providing default value as suppliers in enums

* Fixed test

* Add a supplier for default value to ensure it's calculated correctly

* Added checks for not specifying a body in PUT requests

* Remove synonyms changes

* Remove some supplier changes

* Better call enumParam with supplier version

* Fix compiler error on supplier

* Apply validators or requires depending on index version

* Solved BWC tests that involved using validators instead of requiresParameters

* Add tests

* Spotless

* Update docs/changelog/98268.yaml

* Update changelog

* Update docs/changelog/98268.yaml

* PR comments

* PR feedback

* Serialize index only for new index versions

---------

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

thecoop committed:
Refactor some more uses of inferred versions to explicitly specify index versions (#98588)

kderusso committed:
Query rules docs clarification (#98605)

* Query rules docs clarification

* Update docs/reference/search/search-your-data/search-using-query-rules.asciidoc

* Update docs/reference/search/search-your-data/search-using-query-rules.asciidoc

craigtaverner committed:
Simple ESQL pow() docs fixes after re-reviewing (#98601)

nik9000 committed:
ESQL: LTRIM, RTRIM and fix unicode whitespace (#98590)

Here we add support for the following two ESQL functions:
* LTRIM: remove leading spaces from a string
* RTRIM: remove trailing spaces from a string

We also fix an issue with the handling of unicode white spaces. We
make use of unicode code points to identify unicode whitespace
characters instead of relying on ASCII codes.

Moreover, iterating bytes in a Unicode string needs to consider
that some Unicode characters are encoded using multiple bytes.

rseldner committed:
Add deprecated note for `balanced` allocator (#98610)

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

volodk85 committed:
Bump versions after 8.9.1 release

volodk85 committed:
Prune changelogs after 8.9.1 release

brianseeders committed:
[buildkite] Send Slack notifications to #lucene for lucene_snapshot failures

nik9000 committed:
ESQL: COALESCE function (#98542)

This adds a `COALESCE` function that returns the first non-null value.

benwtrent committed:
Fixing dense vectors bwc tests to reflect new default version (#98621)

New defaults were NOT added for 8.10. They were added for 8.11.

closes: https://github.com/elastic/elasticsearch/issues/98611

relates: https://github.com/elastic/elasticsearch/pull/98268

relates: https://github.com/elastic/elasticsearch/pull/97092

benwtrent committed:
Muting test (#98627)

relates: https://github.com/elastic/elasticsearch/issues/98626

dnhatn committed:
Avoid exceeded semaphore permits in ESQL tests (#98625)

Closes #98603

dnhatn committed:
Query ESQL usage stats from new nodes only (#98617)

The new ESQL stats action doesn't exist on old nodes; 
so we should avoid sending stats-requests to those nodes.

Closes #98604

DaveCTurner committed:
Simplify empty response handler (#98592)

DaveCTurner committed:
Add timeout to AbstractEsqlIntegTestCase#run (#98593)

If this action hangs then the test suite takes a very long time to time
out. This commit imposes a 30s timeout, after which the test fails.

DaveCTurner committed:
Extract ESQL_TEST_EXECUTOR constants (#98594)

Just clears up some non-coincidentally-equal strings which makes it a
bit easier to navigate to the usage sites.

DaveCTurner committed:
Fix assertion in testReturnsImmediatelyIfVoteAlreadyWithdrawn (#98570)

This test is today asserting that the _initial_ state has all the
expected exclusions, but it should be checking the _final_ state.

Relates #98386

DaveCTurner committed:
`Setting` should not catch `AssertionError` (#98543)

We shouldn't be catching `AssertionError` in the production code here,
because that will suppress certain kinds of test failure. Instead we
should allow it to bubble up and fail the test.

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

elasticsearchmachine commented:
Hi @csoulios, I've created a changelog YAML for you.

csoulios committed:
Update docs/changelog/98411.yaml

csoulios committed:
Merge branch 'main' into no-doc-values

csoulios commented:
Closing this as duplicate of #98639
